### ISSUE TITLE
horizontal DateTime ticks overlap

### ISSUE THREAD

**Issue by swharden** [ScottPlot/ScottPlot#725]
There's also a long-running problem where the last tick can flow off the figure. I wonder if there's an easy solution for this...

![image](https://user-images.githubusercontent.com/4165489/106370064-cc26cf00-6324-11eb-859a-67d0e74b4eeb.png)

**Comment by bclehmann** [comment]
# Bounding Rectangle

For the overlapping labels you can use [this overload](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.graphics.drawstring?redirectedfrom=MSDN&view=dotnet-plat-ext-5.0#System_Drawing_Graphics_DrawString_System_String_System_Drawing_Font_System_Drawing_Brush_System_Drawing_RectangleF_System_Drawing_StringFormat_) to specify a bounding rectangle, and then you can decide to truncate or wrap or whatever with [this](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.stringtrimming?view=dotnet-plat-ext-5.0) enum.

# Short Form Dates

An other option is to detect if it goes too wide and then shorten it yourself, perhaps something like July 22, 1989 could become 7/22/89 (yes I am aware this introduces debate on how to show short form dates, however this should be taken care of if the user specifies a specific locale).

For example, if I set the locale to en-us, both the long and short form dates should be correct:
```cs
DateTime day= new DateTime(1989, 7, 22);

// Full form
Console.WriteLine(day.ToString("F", new System.Globalization.CultureInfo("en-us"))); // Saturday, July 22, 1989 12:00:00 AM
Console.WriteLine(day.ToString("F", new System.Globalization.CultureInfo("fr-fr"))); // samedi 22 juillet 1989 00:00:00

// Month-Day form (no year, no time)
Console.WriteLine(day.ToString("M", new System.Globalization.CultureInfo("en-us"))); // July 22
Console.WriteLine(day.ToString("M", new System.Globalization.CultureInfo("fr-fr"))); // 22 juillet

// Short form (year and time)
Console.WriteLine(day.ToString("g", new System.Globalization.CultureInfo("en-us"))); // 7/22/89 12:00 AM
Console.WriteLine(day.ToString("g", new System.Globalization.CultureInfo("fr-fr"))); // 22/07/1989 00:00
```

It is still unpredictable how long the short form will be, e.g. `7/22/89` and `22 07 1989` are different lengths but both short-form dates. And if someone passed in a custom culture they may not think to override the short-form date, which means they get the default (.NET calls it the invariant culture, it's a generic English-language locale).

Plus I don't know how short-form dates generalize to non-gregorian calendars or under what circumstances .NET would use a non-gregorian calendar, but this line prints using the Persian calendar (the locale is Farsi-Iran):

```cs
Console.WriteLine(day.ToString("g", new System.Globalization.CultureInfo("fa-ir"))); // 1368/4/31 0:00
```

It appears both long and short form dates in this locale use the Persian calendar.

Another consideration is when one label might overflow but not the others, in this case, they should all switch to short form at once to prevent confusion. Although none of this addresses the concern that short form dates (especially those with times attached) can still overflow if you zoom far enough out.

**Comment by swharden** [comment]
Hey @bclehmann, this is all excellent information, and changing the string format of DateTime tick labels should be considered as a way to refine the density of ticks to be visually appealing.

The screenshot shows the words "September" and "October" overlapping, so it seems the tick density (based on the measurements of the largest tick label) is being sub-optimally calculated. I suspect the cause of this problem is in one of these two areas:

### Idea 1: multi-pass label size

Perhaps the first-pass tick labels are small (year) but second-pass tick labels are larger (month). The density is calculated on the largest measured label from the first pass. I'm not sure how to best solve this, because this is a recursive problem (changing density may change tick label content, requiring a new change in density).

https://github.com/ScottPlot/ScottPlot/blob/486f2ae38b9a4edb8eeeb1e8bc92c5e1a425ffca/src/ScottPlot/Ticks/TickCollection.cs#L50-L64

### Idea 2: improve density calculation

Perhaps there's a simpler issue in the form of a bug that could be resolved by looking more closely at the method that calculates the density for DateTime ticks given maximum tick label size and the dimensions of the plot.

https://github.com/ScottPlot/ScottPlot/blob/486f2ae38b9a4edb8eeeb1e8bc92c5e1a425ffca/src/ScottPlot/Ticks/TickCollection.cs#L111-L162

**Comment by bclehmann** [comment]
# Simple solution:

Only two and a half passes, you run it once with the constant max widths, then run it again with the new max widths but with the same `DateTimeUnit`. In other words, you avoid the recursion by only changing the density, but changes in density do not cause changes in label size. I don't know how good this solution would look, it might give labels whose ticks are too sparse.

# Sidestep Solution:

The problem is exceedingly unlikely to happen with rotated tick labels as they take up less space horizontally than they reserve.

# Multi-Pass

**Disclaimer:** This algorithm might not actually work because I've only tried it in my head.

I feel pretty bad about suggesting an algorithm without implementing it (or even fully convincing myself that it's correct) but I'm a little busy right now. I might try my hand at an implementation but I may be a while. Don't feel obliged to try and get my disconnected thoughts working.

Yes this is recursive, but it also can be written so it's guaranteed to halt. If changing the density _does not_ change tick label content then you can simply terminate (e.g. the change in density is too small to trigger different content). It cannot expand the label without choosing a more precise unit, and it can never shrink the label. And it trivially halts once you get to a point where you cannot make the label any wider (e.g. July 22, 1989 15:30.000) is as precise as ScottPlot allows on a `DateTimeAxis` which means that once you reach milliseconds it will always give the same labels. Also note that most of the time it will just tweak the density by a minor amount which is unlikely to cascade into a million reflows.

So then your recursive function becomes this:

1. Get labels and positions given constant max widths
2. Use the length of the longest label to run your next pass
3. If the label widths increase you run another pass with the new max label sizes
    3.1 If not: terminate. Note that this base case is guaranteed to occur because ScottPlot does not have a `DateTimeTickUnit` smaller than milliseconds

The unfortunate part is that this may zoom all the way down to milliseconds with an axis with one ticklabel. That's because this is a greedy algorithm which reserves more space until it either gets satisfied _or_ it decides it's going to take up more space with it's newly found free space. There are a few solutions to this:

- Cap how many times it can reserve more space, either through recursion depth or only allowing it to travel through certain `DateTimeTickUnits`
- Require that it always has a "sensible" number of axis labels, e.g. if it gives a solution with < 5 axis labels you halt and use the one before, altering it to make it valid if necessary
- Observe that almost all solutions which were invalid would become valid if you imposed a less precise `DateTimeUnit`
    - You can treat your call stack (or any other way to store your different label contents and positions) upon termination as a menu of solutions which may be valid if you ran them again with the same `DateTimeUnit` as their caller used.

You'll notice that most of the ways to tame this algorithm are putting a floor on the number of axis labels that a valid solution must have or a ceiling on the number of `DateTimeUnit`s that you want to let the algorithm traverse through.

**Comment by bclehmann** [comment]
I implemented the multi-pass approach: https://github.com/bclehmann/ScottPlot/blob/972b15efaa5cac49af318b39f01616bd9430010f/src/ScottPlot/Ticks/TickCollection.cs

It was slightly hairier than I thought, it is possible for it to become periodic. I combated this with a set, and I believe it should always have a periodicity of 2 (or 1, which is what I originally thought would happen). So I wait for it to repeat twice and then I break out of the loop. Then I look at the two most recent solutions for the one that gives the most space to the ticklabels and use that.

It seems to work, but it is the sort of thing that's hard to validate either by a proof or by testing, at least for me.

One issue unrelated to the overlapping, is that it doesn't seem "stable", for very precise zoom levels simply panning can cause it to change the number of tick labels causing it to flicker some of them when panning quickly. This is because the level of zoom is so precise that when panning causes a new tick to come into view (or especially one with a longer label) it needs to remove ticks to prevent overlapping. I think any algorithm that attempts to find an optimal solution would do this, but it isn't a nice visual effect. Perhaps padding can be added after to prevent this?

**Comment by bclehmann** [comment]
It looks like adding padding like this can alleviate the flickering but to completely eliminate it (especially at very high zoom on very long axis labels such as dates) it requires intractably large padding which means even when no padding is required the labels are very far apart.

```cs
candidate.width += candidate.width % 5;
candidate.height += candidate.height % 5;
```

**Comment by bclehmann** [comment]
If you're concerned about massive recursion depth it seems (experimentally) that it never makes more than 6 passes (not including the first pass), with more than 4 being very rare. And it's written iteratively as well, so it doesn't tank the callstack.

**Comment by swharden** [comment]
> I implemented the multi-pass approach: https://github.com/bclehmann/ScottPlot/blob/overlap-datetime/src/ScottPlot/Ticks/TickCollection.cs

Can you throw in a stopwatch to benchmark how long that loop takes with/without this feature? I'm a little weary of making tons of calls to GDI+'s MeasureString() method... It wouldn't surprise me if font rendering (especially with anti-aliasing) is slower than expected, and I don't really know how GDI+ measures strings under the hood. Having some actual numbers might confirm or eliminate the concern of having tons of strings measured on every render.

> It was slightly hairier than I thought

This is the story of my life every time I open up the tick module 🤪 

> massive recursion depth

**How's this for a solution?** I thought of it today and it seems like a solution half way between a hack and something reasonable:

1.) Assume the month format is the one with the widest potential size*
2.) Therefore, measure the width of the widest month label (irrespective of what months are plotted)
3.) Calculate tick density based on that

_*millisecond-precision time codes might be longer, so perhaps only apply this shortcut if/when the axis span makes sense for it to be used. Perhaps this rule could apply only whenever the axis spans more than 1 day._

This solution would never result in overlapping labels, and its only downside is that shorter ticks (perhaps just the year) would appear to have a little extra space between them, but I don't think it would be visually unappealing. If anything, it would ensure tick density is always approximately the same.

**Comment by bclehmann** [comment]
> Can you throw in a stopwatch to benchmark how long that loop takes with/without this feature?

Sure thing, I put a timer around the loop (not the whole function) and I got this:


--| --
-- | --
N | 5791
Min (s) | 0.0000719
Max (s) | 0.0043229
Mean (s) | 0.000211
Median (s) | 0.000181
StdDev (s) | 0.0002365



A histogram (with incomprehensible axis labels because excel):
![image](https://user-images.githubusercontent.com/8635304/107887257-1336c780-6ec2-11eb-9ab7-0e3a7d90d50f.png)

**Comment by bclehmann** [comment]
As for the assuming the month is the longest, I thought this would be tricky to internationalize, but it turns out one can just use this: `DateTimeFormatInfo.MonthGenitiveNames` https://docs.microsoft.com/en-us/dotnet/api/system.globalization.datetimeformatinfo.monthgenitivenames?view=net-5.0#System_Globalization_DateTimeFormatInfo_MonthGenitiveNames

One can loop over this and get the longest month in the given culture, including in non-gregorian calendars. Note that you can't use string length, you'd have to use `MeasureString`, as some character's are wider than others in a given font. Fortunately, the length of this array is **always** 13 (even in calendars with less than 13 months), so it's not like it should cause a big issue.

The difference between genitive and non-genitive month names is a little confusing, in some languages (in the docs they list Russian) a month may be written differently when written as part of a date (genitive case) versus written outside of a date (non genitive). So for our purposes, we would want the genitive month names.

I think this approach would look alright, I don't think it'd look too weird to have the shorter month names spaced out, unless of course there's a language with a really short month name as well as one that's 40 characters.

**Comment by swharden** [comment]
I fixed this in #925 using the "longest month" strategy above. I want to put a lot more work into the tick module and I added that to the triage list.

Before|After
---|---
![image](https://user-images.githubusercontent.com/4165489/112909364-45416900-90bf-11eb-84b7-53222643588c.png)|![image](https://user-images.githubusercontent.com/4165489/112914434-f4833d80-90c9-11eb-8dee-ad29bd002751.png)
